internal/daemon/daemon.go

```diff
@@ -422,22 +433,25 @@ func (d *Daemon) StartAPI(bootstrap bool, initConfig map[string]string, newConfi
 		return err
 	}
 
-	// Extract user defined endpoints for core listener.
-	coreEndpoints, err := resources.GetAndValidateCoreEndpoints(d.extensionServers)
+	// Validate the extension servers again now that we have applied addresses.
+	err = resources.ValidateEndpoints(d.extensionServers, d.address.URL.Host)
 	if err != nil {
 		return err
 	}
 
-	serverEndpoints := []rest.Resources{resources.InternalEndpoints, resources.PublicEndpoints}
-	serverEndpoints = append(serverEndpoints, coreEndpoints...)
-	server := d.initServer(serverEndpoints...)
-	network := endpoints.NewNetwork(d.shutdownCtx, endpoints.EndpointNetwork, server, d.address, d.ClusterCert())
 	err = d.endpoints.Down(endpoints.EndpointNetwork)
 	if err != nil {
 		return err
 	}
 
-	err = d.endpoints.Add(network)
+	serverEndpoints := []rest.Resources{resources.InternalEndpoints, resources.PublicEndpoints}
+	err = d.addCoreServers(false, d.address, d.ClusterCert(), serverEndpoints)
+	if err != nil {
+		return err
+	}
+
+	// Add extension servers before post-join hook.
+	err = d.addExtensionServers(false, d.ClusterCert(), d.address.URL.Host)
```

Comment:
nice, this makes a lot of sense since additional network listeners should be added regardless if we are bootstrapping or joining the cluster!